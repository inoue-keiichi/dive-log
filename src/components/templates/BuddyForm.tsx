import { Buddy, buddySchema } from "@/schemas/buudy";
import { DiveLog } from "@/schemas/diveLog";
import { ResponseError } from "@/utils/type";
import { zodResolver } from "@hookform/resolvers/zod";
import {
  Button,
  FormControl,
  FormHelperText,
  InputLabel,
  OutlinedInput,
  Stack,
  Typography,
} from "@mui/material";
import { FC, useState } from "react";
import { FieldErrors, useForm } from "react-hook-form";

const hasInvalidField = (errors: FieldErrors<DiveLog>) => {
  return Object.values(errors).filter((v) => v.message).length > 0;
};

type Props = {
  onSubmit: (buddy: Buddy) => Promise<void>;
  error?: ResponseError;
};

const BuddyForm: FC<Props> = (props) => {
  const { onSubmit, error } = props;

  const [loading, setLoading] = useState(false);

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<Buddy>({
    resolver: zodResolver(buddySchema),
  });

  return (
    <Stack
      spacing={2}
      sx={{
        width: "100%",
        backgroundColor: "white",
        padding: "20px 40px 20px 40px",
        borderRadius: 2,
      }}
      component="form"
      onSubmit={handleSubmit((buddy) => {
        setLoading(true);
        onSubmit(buddy).finally(() => {
          setLoading(false);
        });
      })}
    >
      <Typography>あなたの名前を入力してください</Typography>
      <FormControl>
        <InputLabel htmlFor="name">名前</InputLabel>
        <OutlinedInput
          id="name"
          label="名前"
          type="text"
          error={!!errors.name?.message}
          {...register("name")}
        />
        <FormHelperText error={!!errors.name?.message}>
          {errors.name?.message ?? ""}
        </FormHelperText>
      </FormControl>
      <FormHelperText error={!!error}>{error?.message ?? ""}</FormHelperText>
      <Button
        sx={{ width: "30%" }}
        disabled={!!error || hasInvalidField(errors)}
        variant="contained"
        type="submit"
        loading={loading}
        loadingPosition="end"
      >
        次へ
      </Button>
    </Stack>
  );
};

export default BuddyForm;
