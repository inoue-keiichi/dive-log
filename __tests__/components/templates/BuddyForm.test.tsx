import BuddyForm from "@/components/templates/BuddyForm";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { fill } from "../__utils__/common";

const submit = jest.fn();
afterEach(() => {
  submit.mockClear();
});

describe("BuddyForm", () => {
  it("ssucceeds in submitting a buddy", async () => {
    render(<BuddyForm onSubmit={submit} />);
    fill(screen.getByLabelText("名前"), "JAI");
    const submitButton = screen.getByText("次へ");
    fireEvent.click(submitButton);
    // react-hook-form によって submit が呼び出されるまで待機
    await waitFor(() => expect(submit).toHaveBeenCalledTimes(1));
  });

  it("fails to submit a buddy when name is empty", async () => {
    render(<BuddyForm onSubmit={submit} />);
    const submitButton = screen.getByText("次へ");
    fireEvent.click(submitButton);
    // react-hook-form によって submit が呼び出されるまで待機
    await waitFor(() => expect(submit).toHaveBeenCalledTimes(0));
  });

  it("snapshot", async () => {
    const tree = render(<BuddyForm onSubmit={submit} />);
    expect(tree).toMatchSnapshot();
  });
});
