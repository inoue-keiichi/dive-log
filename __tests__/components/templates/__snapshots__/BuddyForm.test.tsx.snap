// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuddyForm snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="MuiStack-root css-1ghstqc-MuiStack-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
        >
          あなたの名前を入力してください
        </p>
        <div
          class="MuiFormControl-root css-1iw3t7y-MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="name"
          >
            名前
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
              id="name"
              name="name"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-81qg8w"
              >
                <span>
                  名前
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-er619e-MuiFormHelperText-root"
          />
        </div>
        <p
          class="MuiFormHelperText-root css-18y2hez-MuiFormHelperText-root"
        />
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-14ammhu-MuiButtonBase-root-MuiButton-root"
          id=":r2:"
          tabindex="0"
          type="submit"
        >
          次へ
          <span
            class="MuiButton-loadingWrapper"
            style="display: contents;"
          />
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="MuiStack-root css-1ghstqc-MuiStack-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      >
        あなたの名前を入力してください
      </p>
      <div
        class="MuiFormControl-root css-1iw3t7y-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="name"
        >
          名前
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
            id="name"
            name="name"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-81qg8w"
            >
              <span>
                名前
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-er619e-MuiFormHelperText-root"
        />
      </div>
      <p
        class="MuiFormHelperText-root css-18y2hez-MuiFormHelperText-root"
      />
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-14ammhu-MuiButtonBase-root-MuiButton-root"
        id=":r2:"
        tabindex="0"
        type="submit"
      >
        次へ
        <span
          class="MuiButton-loadingWrapper"
          style="display: contents;"
        />
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
